### PR TITLE
better handle of flow closeables

### PR DESCRIPTION
**Pull request by talIguaz**

### CODE REVIEW

File: storey/flow.py [hunk 1]
```diff
@@ -194,6 +194,8 @@ def __init__(self, initial_state, fn, group_by_key=False, **kwargs):
         self._state = initial_state
         self._fn = fn
         self._group_by_key = group_by_key
+        if hasattr(initial_state, 'close'):
+            self._closeables = [initial_state]
```
Comment (gtopper) [inline storey/flow.py:198]: Shouldn't this be appended instead?

File: storey/sources.py [hunk 2]
```diff
@@ -115,15 +115,17 @@ async def _run_loop(self):
...
             except BaseException as ex:
                 self._ex = ex
                 if not self._q.empty():
                     self._q.get()
                 self._termination_future.set_result(None)
                 break
+            finally:
+                if event is _termination_obj or self._ex:
+                    for closeable in self._closeables:
+                        await closeable.close()
```
Comment (gtopper) [inline storey/sources.py]: Hm, just put this after the loop?

File: storey/sources.py [hunk 3]
```diff
@@ -261,6 +261,10 @@ async def _run_loop(self):
                 if not self._q.empty():
                     await self._q.get()
                 return None
+            finally:
+                if event is _termination_obj or self._ex:
+                    for closeable in self._closeables:
+                        await closeable.close()
```
Comment (gtopper) [inline storey/sources.py:267]: Hm, just put this after the loop?

File: storey/flow.py (hunk 1 above)
Comment (talIguaz) [reply storey/flow.py:198]: as we discussed, each step sets his `_closeables` list and the Flow.run() is responsible on appending all the closeables into one list

Comment (gtopper) [reply storey/flow.py:198]: Ok, it's just that we already create a list in Flow.__init__ 🤷‍♂️

**Review by gtopper — APPROVED** [review]

**Pull request merged**